signalapp/Signal-iOS
# Fixes strange alignments of the selection views in a chat

### Contributor checklist

 * iPhone 11, iOS 13.5
 * iPhone 12, iOS 14.2

- - - - - - - - - -

### Description


The bug was originally reported in #4552 . This strange behavior is was due to a bug in UIStackView that behaves strangely when its subviews are hidden and then shown. The existing code has worked around the bug by adding `alpha` values along with setting `hidden` attribute. However the workaround sometimes doesn't work when the subviews of a stack view is hidden and shown, and the stack view also has a stretching subview so the placement of the subviews also changes when a subview is hidden and is shown.

The code of the PR to fix the bug adds and removes the problematic subviews instead of just hiding and showing them to better work around the `UIStackView` bug.

Note that a similar issue is fixed in #4636 by using the `hide by alpha=0` trick. It works for that PR because it does not involve stretching subviews.

### Steps to verify the fix
It's probably easier to see the gif below to understand the steps.
1. Have a chat long enough so you must scroll to reach the bottom of it
    * have sent messages, received messages and system messages
2. Scroll up the chat so a message is the next one to appear on screen
3. Go back to the conversation list screen
4. Come back to the chat again
5. Scroll down to the bottom

**Actual behavior before the fix**

The checkbox (`selectionView`) appears at misaligned for the message. For the system messages, the checkbox appears even when the view is *not* in selection mode.

**Expected behavior after the fix**

The checkbox (`selectionView`) appears aligned. For the system messages, the checkbox does not appear when the view is not in selection mode.

### Screenshots

|Before|After|
|---|---|
|![checkbox-bug-1](https://user-images.githubusercontent.com/28482/100289333-a48f4100-2f46-11eb-953e-1056cd2c9024.png)|![checkbox-fixed-1](https://user-images.githubusercontent.com/28482/100289426-dd2f1a80-2f46-11eb-8264-0a94423bd110.png)|

I'll paste more screenshots and gifs in the comments.
